syntax = "proto3";
package minecpp.chunk_storage;

import "chunk.proto";

service ChunkStorage {
    rpc LoadChunk (LoadChunkRequest) returns (minecpp.chunk.NetChunk);
    rpc SetBlock (SetBlockRequest) returns (EmptyResponse);
    rpc AddReferences (AddReferencesRequest) returns (AddReferencesResponse);
    rpc RemoveReference (RemoveReferencesRequest) returns (EmptyResponse);
    rpc HeightAt (HeightAtRequest) returns (HeightAtResponse);
}

message LoadChunkRequest {
    int32 x = 1;
    int32 z = 2;
}

message SetBlockRequest {
    int32 x = 1;
    int32 y = 2;
    int32 z = 3;
    int32 state = 4;
}

message Coord {
    int32 x = 1;
    int32 z = 2;
}

message AddReferencesRequest {
    bytes engine_id = 1;
    bytes player_id = 2;
    repeated Coord coords = 3;
}

enum ReferenceStatus {
    OK = 0;
    MUST_MOVE = 1;
}

message AddReferencesResponse {
    ReferenceStatus status = 1;
    bytes target_engine_id = 2;
}

message RemoveReferencesRequest {
    bytes player_id = 1;
    repeated Coord coords = 2;
}

message HeightAtRequest {
    int32 x = 1;
    int32 z = 2;
}

message HeightAtResponse {
    int32 height = 1;
}

message EmptyResponse {
}

